src/nvim/fileio.c

```diff
@@ -1883,16 +1883,20 @@ readfile (
       xfree(keep_msg);
       keep_msg = NULL;
       msg_scrolled_ign = TRUE;
-      p = msg_trunc_attr(IObuff, FALSE, 0);
+
+      if (!read_stdin && !read_buffer) {
+        p = msg_trunc_attr(IObuff, FALSE, 0);
```

Comment:
Previously `p` was always set here. Now `p` can (in the else-branch) point to something freed and `set_keep_msg(p,0)` may be called with this pointer value. Should `p` be set to `NULL` in the else-branch?

```
==5448==ERROR: AddressSanitizer: heap-use-after-free on address 0x631000050884 at pc 0x00000046c232 bp 0x7ffe3fdb65e0 sp 0x7ffe3fdb5d90
READ of size 1 at 0x631000050884 thread T0
    #0 0x46c231 in __interceptor_strlen.part.31 /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:284
    #1 0xf0c706 in xstrdup /home/oni-link/git/neovim/src/nvim/memory.c:401:24
    #2 0x16e1bbd in vim_strsave /home/oni-link/git/neovim/src/nvim/strings.c:51:20
    #3 0xf29e1e in set_keep_msg /home/oni-link/git/neovim/src/nvim/message.c:1017:16
    #4 0xc79d95 in readfile /home/oni-link/git/neovim/src/nvim/fileio.c:1898:9
    #5 0x655f3e in open_buffer /home/oni-link/git/neovim/src/nvim/buffer.c:176:16
    #6 0xdf01c1 in read_stdin /home/oni-link/git/neovim/src/nvim/main.c:1337:9
    #7 0xde1d6d in main /home/oni-link/git/neovim/src/nvim/main.c:418:5
    #8 0x7f765c839730 in __libc_start_main (/lib64/libc.so.6+0x20730)
    #9 0x4473b8 in _start (/home/oni-link/git/neovim/build/bin/nvim+0x4473b8)

0x631000050884 is located 132 bytes inside of 65537-byte region [0x631000050800,0x631000060801)
freed by thread T0 here:
    #0 0x4fcb20 in __interceptor_cfree.localalias.1 /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:54
    #1 0xf0b834 in xfree /home/oni-link/git/neovim/src/nvim/memory.c:130:3
    #2 0xc6ffe8 in readfile /home/oni-link/git/neovim/src/nvim/fileio.c:1012:9
    #3 0x655f3e in open_buffer /home/oni-link/git/neovim/src/nvim/buffer.c:176:16
    #4 0xdf01c1 in read_stdin /home/oni-link/git/neovim/src/nvim/main.c:1337:9
    #5 0xde1d6d in main /home/oni-link/git/neovim/src/nvim/main.c:418:5
    #6 0x7f765c839730 in __libc_start_main (/lib64/libc.so.6+0x20730)

previously allocated by thread T0 here:
    #0 0x4fccd8 in malloc /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:64
    #1 0xf0b594 in try_malloc /home/oni-link/git/neovim/src/nvim/memory.c:84:15
    #2 0xf0b5d4 in verbose_try_malloc /home/oni-link/git/neovim/src/nvim/memory.c:100:15
    #3 0xc6fe7f in readfile /home/oni-link/git/neovim/src/nvim/fileio.c:1001:24
    #4 0x655f3e in open_buffer /home/oni-link/git/neovim/src/nvim/buffer.c:176:16
    #5 0xdf01c1 in read_stdin /home/oni-link/git/neovim/src/nvim/main.c:1337:9
    #6 0xde1d6d in main /home/oni-link/git/neovim/src/nvim/main.c:418:5
    #7 0x7f765c839730 in __libc_start_main (/lib64/libc.so.6+0x20730)

SUMMARY: AddressSanitizer: heap-use-after-free /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:284 in __interceptor_strlen.part.31

```